### PR TITLE
Simplify build_docs command to avoid patching code from Sphinx

### PR DESCRIPTION
**Pull request by astrofrog**
Prior to this, we were live-patching code from Sphinx, which seems very brittle. The subprocess code now uses public Sphinx API to do its thing.

This appears to solve https://github.com/astropy/sphinx-astropy/issues/12 - the issue that the docs were currently being built in ``docs/docs/_build/html``.

At the moment this breaks the ``no_intersphinx`` feature because it's not trivial to do this via the ``sphinx-build`` API. I'm curious whether people use this?

I also need to add back to creating of ``_static`` if it's actually needed (not sure if it is anymore)

### CODE REVIEW

File: astropy_helpers/commands/build_sphinx.py [hunk 1]
```diff
@@ -162,51 +153,14 @@ def run(self):
...
-        if self.no_intersphinx:
-            # the confoverrides variable in sphinx.setup_command.BuildDoc can
-            # be used to override the conf.py ... but this could well break
-            # if future versions of sphinx change the internals of BuildDoc,
-            # so remain vigilant!
-            subproccode = subproccode.replace(
-                'confoverrides = {}',
-                'confoverrides = {\'intersphinx_mapping\':{}}')
+        subproccode += textwrap.dedent("""
+            argv = ['-W', '-b', 'html', '.', '_build/html']
```
Comment (pllim) [inline astropy_helpers/commands/build_sphinx.py]: This hardcodes `_build` though. What if there is a package out there that uses helpers but not `_build`?

Comment (astrofrog) [reply astropy_helpers/commands/build_sphinx.py]: I agree, I've fixed this now

File: astropy_helpers/commands/build_sphinx.py [hunk 2]
```diff
@@ -162,51 +152,16 @@ def run(self):
...
+        subproccode += textwrap.dedent("""
+
+            if SPHINX_LT_17:
+                argv.insert(0, 'sphinx-build')
+
+            for builder in {builders!r}:
+                argv = ['-W', '.', '-b', builder, os.path.join({output_dir!r}, builder)]
+                build_main(argv=argv)
+        """.format(builders=self.builder,
+                   output_dir=os.path.relpath(self.build_dir, 'docs')))
```
Comment (pllim) [inline astropy_helpers/commands/build_sphinx.py]: Better but not quite there as "docs" is still hardcoded. There are projects with say, "doc", instead of "docs".

Comment (astrofrog) [reply astropy_helpers/commands/build_sphinx.py]: I think that should be fixed now

File: astropy_helpers/commands/build_sphinx.py [hunk 3]
```diff
@@ -162,51 +148,16 @@ def run(self):
...
+        subproccode += textwrap.dedent("""
+
+            if SPHINX_LT_17:
+                argv.insert(0, 'sphinx-build')
+
+            for builder in {builders!r}:
+                argv = ['-W', '.', '-b', builder, os.path.join({output_dir!r}, builder)]
+                build_main(argv=argv)
+        """.format(builders=self.builder,
+                   output_dir=os.path.abspath(self.build_dir)))
```
Comment (pllim) [inline astropy_helpers/commands/build_sphinx.py]: As long as Sphinx does not impose path length limit, this should be okay.

**Comment by saimn** [comment]
What about the debug, verbosity, and traceback options ? These can be useful for debugging, or to avoid lengthy output as you are using `-v` by default.

**Comment by astrofrog** [comment]
@saimn - I think those options are still on the base class for the command:

https://github.com/sphinx-doc/sphinx/blob/master/sphinx/setup_command.py

so I think they can still be used?

**Comment by saimn** [comment]
Hmm I can see pdb there, but how is it passed to sphinx's `build_main` ? And as you are hard-coding `-v` it will always be in verbose mode. I may be missing something ;)

**Comment by astrofrog** [comment]
@saimn - ah right yes sorry I am being dumb. I'll fix it up.

**Comment by astrofrog** [comment]
This now also gets rid of the custom treatment of the warnings as errors - Sphinx has infrastructure to do this, and furthermore the default behavior is to do a recap of the warnings at the end of the build (@Cadair pointed this out to me before):

```
  RemovedInSphinx20Warning)
[automodsumm] index.rst: found 5 automodsumm entries to generate
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 5 source files that are out of date
updating environment: 10 added, 0 changed, 0 removed
reading sources... [100%] noncelestial                                         
looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
writing output... [100%] noncelestial                                          

Warning, treated as error:
/Users/tom/Dropbox/Code/Astropy/reproject/docs/index.rst:8:py:obj reference target not found: erere
```

so I think this is an improvement!

Once https://github.com/astropy/sphinx-astropy/pull/13 is merged I'll also add back the option for the intersphinx once https://github.com/astropy/sphinx-astropy/pull/13 is merged and a new version of sphinx-astropy is released.

**Comment by Cadair** [comment]
@pllim I ran it on sunpy and it broke! :tada:

**Comment by astrofrog** [comment]
This is ready for review/merging. The code is only a bit shorter than before but should be a lot clearer and doesn't include any funny business with patching code from Sphinx dynamically.

The intersphinx option will only work once sphinx-astropy 1.1 is released, but it'll be easy to release that version before astropy 3.1 is out (and there is a nice warning if the version of sphinx-astropy is too old)

**Comment by pllim** [comment]
tl;dr -- Does this introduce any behavioral change; i.e., `python setup.py build_docs` or `python setup.py build_sphinx` for people using helpers will continue to work the same?

@Cadair , you should test this again, since you reported failures earlier.

**Comment by astrofrog** [comment]
@pllim - no procedural change - everything is exactly as it was before in terms of command-line API.

File: astropy_helpers/commands/build_sphinx.py [hunk 4]
```diff
@@ -66,19 +117,23 @@ def initialize_options(self):
         self.no_intersphinx = False
         self.open_docs_in_browser = False
         self.warnings_returncode = False
+        self.traceback = False
```
[the remark this answers is not in the document]
Comment (astrofrog) [reply astropy_helpers/commands/build_sphinx.py]: It is inherited from ``BuildDocs``, and I'm changing the default to match the behavior we've had so far.

File: astropy_helpers/commands/build_sphinx.py [hunk 5]
```diff
@@ -1,23 +1,78 @@
...
 import glob
-import textwrap
 import warnings
+from distutils.version import LooseVersion
 
 from distutils import log
-from distutils.cmd import DistutilsOptionError
 
-import sphinx
+from sphinx import __version__
```
[the remark this answers is not in the document]
Comment (astrofrog) [reply astropy_helpers/commands/build_sphinx.py]: Fixed - I did it myself to make sure I changed all instances at the same time.

File: astropy_helpers/commands/build_sphinx.py [hunk 6]
```diff
@@ -1,23 +1,78 @@
...
 from sphinx.setup_command import BuildDoc as SphinxBuildDoc
 
-from ..utils import minversion, AstropyDeprecationWarning
+from ..utils import AstropyDeprecationWarning
+
+SPHINX_LT_17 = LooseVersion(__version__) < LooseVersion('1.7')
+
+SUBPROCESS_CODE = """
```
[the remark this answers is not in the document]
Comment (astrofrog) [reply astropy_helpers/commands/build_sphinx.py]: Fixed - I did it myself to make sure I changed all instances at the same time.

File: astropy_helpers/commands/build_sphinx.py [hunk 7]
```diff
@@ -26,10 +81,6 @@ class AstropyBuildDocs(SphinxBuildDoc):
     that is built by the setup ``build`` command, rather than whatever is
     installed on the system.  To build docs against the installed version, run
     ``make html`` in the ``astropy/docs`` directory.
-
-    This also automatically creates the docs/_static directories--this is
```
[the remark this answers is not in the document]
Comment (astrofrog) [reply astropy_helpers/commands/build_sphinx.py]: You only get a warning from Sphinx if you define ``html_static_path`` but don't use it.

The package template, and by extension a number of affiliated packages don't define ``html_static_path`` by default, so no warning is emitted.

The core package defines ``html_static_path`` and has files in ``_static``.

Basically the previous code was just to catch the case where people were defining ``html_static_path`` and then not actually using it. I don't really think we should be working around this, since it's an easy fix (just don't define ``html_static_path``).

[the remark this answers is not in the document]
Comment (eteq) [reply astropy_helpers/commands/build_sphinx.py]: While I see your point, this is a subtle incompatibility.  That is, if there's anyone who defined `html_static_path` and counted on this machinery to build their `_static` will now see this breaking their build, and probably not know why (because the point of this stuff was to make it so they don't have to know about it).

So I'd say this should either stay or we should for at least one version create a more helpful warning that tells people exactly what to do.

[the remark this answers is not in the document]
Comment (astrofrog) [reply astropy_helpers/commands/build_sphinx.py]: To make sure we only show the warning for users who both have the setting set and the folder missing, and to have the message at the end of the build, I added this as a mini-extension in sphinx-astropy: https://github.com/astropy/sphinx-astropy/pull/13#issuecomment-438760820

**Comment by bsipocz** [comment]
I'll try to backport this in hope it provides a solution for https://github.com/astropy/astropy-helpers/issues/415.
